src/OrleansEventSourcing/EventStorage/LogViewAdaptor.cs

```diff
@@ -0,0 +1,360 @@
+
+                    successfulread = true;
+
+                    LastPrimaryIssue.Resolve(Host, Services);
+
+                    Services.Log(Severity.Verbose, "Read success from={0} to={1}", response.FromVersion, response.ToVersion);
+                }
+                catch (Exception e)
+                {
+                    LastPrimaryIssue.Record(new LoadEventsFailed() { Exception = e }, Host, Services);
+                }
+
+                if (successfulread)
+                {
+                    // check validity of response
+                    if (response.ToVersion < response.FromVersion || response.Events == null || response.Events.Count != (response.ToVersion - response.FromVersion))
+                    {
+                        throw new OrleansException($"storage returned invalid response");
+                    }
+                    if (response.FromVersion != ConfirmedVersionInternal)
+                    {
+                        throw new OrleansException($"storage returned wrong events, starting at {response.FromVersion} instead of {ConfirmedVersionInternal}.");
+                    }
+                    if (response.ToVersion < ConfirmedVersionInternal)
+                    {
+                        throw new OrleansException($"storage returned stale version {response.ToVersion}, expected at least {ConfirmedVersionInternal}.");
+                    }
+
+                    // construct state from stream of events
+                    foreach (var item in response.Events)
+                    {
+                        if (lookfor.HasValue && lookfor.Value.Equals(item.Key))
+                            guid_was_seen = true;
+
+                        var @event = (TEvent)item.Value;
+
+                        try
+                        {
+                            Host.UpdateView(ConfirmedStateInternal, @event);
+                        }
+                        catch (Exception e)
+                        {
+                            Services.CaughtUserCodeException("UpdateView", nameof(ReadAsyncInternal), e);
+                        }
+
+                        ConfirmedVersionInternal++;
+                    }
+
+                    return guid_was_seen; // exit the loop, we successfully read all events
+                }
+
+                await LastPrimaryIssue.DelayBeforeRetry();
+            }
+        }
+
+        // write entries in pending queue as a batch. 
+        // Retry until outcome has been conclusively determined.
+        // Return 0 if no updates were written, or the size of the whole batch if updates were written.
+        protected override async Task<int> WriteAsync()
+        {
+            var updates = GetCurrentBatchOfUpdates();
+
+            // record one of the guids in the batch, we will use it for idempotence check on failures
+            Guid guidForIdempotenceCheck = updates.First().Guid;
```

Comment:
well, you are assuming that all derived classes will do this correctly, but since `WriteAsync` is a protected method, somebody could call it differently, right?